### ISSUE TITLE
Research: Jekyll image plugins

### ISSUE THREAD

**Issue by mtlynch** [mtlynch/mtlynch.io#116]
The minimal mistakes 4.5.0 theme changed the width of the content area a bit. It used to be 764px wide, so I made 764px width copies of my images and then linked them to the full size image. But 4.5.0 seems to have expanded the width to 984px, so now all of my custom sizing looks weird:

![image](https://user-images.githubusercontent.com/7783288/29157729-1840c690-7d76-11e7-9081-5215e3819bf4.png)

I've seen some discussion of plugins that handle some of this image work for you automatically. A few examples (there may be more):

* https://github.com/wildlyinaccurate/jekyll-responsive-image
* https://github.com/robwierzbowski/jekyll-picture-tag
* https://github.com/netlify/jekyll-srcset

I'm interested in a solution that (in priority order):

* Minimizes the effort of adding an image to a post
* Adapts to changes in the theme or device
* Prevents users from downloading excessive data (e.g. downloading a 3440x1440 px image to be displayed in a 200px space).

**Comment by shredtechular** [comment]
I have compiled some thoughts and outcomes based on some preliminary research so far.  I could probably spend days doing research in this area and playing with the different options, but wanted to time-box this to a couple hours to begin with.  If there are areas you would like me to explore further or have any questions, certainly let me know.  At any rate, here are my thoughts on this issue and my research so far:

## Over-arching issue & solution
These days one of the first ways to make sure you don't end up with weird spacing issues, like in the example that is given in the issue with the large chunk of empty whitespace on the right, is simply using CSS.  Lots of responsive frameworks, like [Bootstrap have this solved using CSS](https://getbootstrap.com/docs/3.3/css/#images-responsive).  

It appears that images that have this problem in the current site are wrapped with the `<a>` tag that makes it possible to have the popup modal when clicking on the image.  Images that are not configured to have the popups look like they do not have this issue that I can see?  So, the first step I would take with any of the below solutions is to add a bit of CSS to expand the image out to 100% responsive either way which will ensure you have full image container width at all browser widths.  Unless you want the image to breakdown at only certain breakpoints and not fully scale responsively?

The CSS solution would not **change the effort of adding an image to a post** and as stated above it certainly will **adapt to changes in the theme and device**.  The CSS will not solve the **Preventing users from downloading excessive data** problem, however.  For that requirement, we do need another solution to minimize which size image is downloaded given the width/device a user is browsing the site on.  A great list was provided in the issue above and I know of a SaaS solution for doing this as well that I also will briefly discuss below:

##  [Jekyll-responsive-image plugin](https://github.com/wildlyinaccurate/jekyll-responsive-image)
### TL;DR Overview: 
Very flexibile and configurable so I don't see any loss of current functionality if we can get this to work in the site.  
### Pros:
- It does solve the requirements outlined.
- very configurable in terms of what size images you want to create, and can create distinct templates to use for images.  For example, I know the site uses `<figure>` tags and captions and we could continue to use this format with this plugin.  
- no loss of current functionality.  For example, I think the popup modal piece we can still have working here with pretty minimal effort.
- Actively supported and releasing new features/fixes.  Last release was in April of 2017.
### Cons:
- Complexity.  I would say this is probably the more complex plugin of the group with a lot of options to tweak/configure.  Because of the complexity it is very flexible as stated above, but it is complex.
- LOE to integrate.  Attempted to integrate this in the site as it stands today and couldn't get it to work.  I was able to integrate using the same configs with a test site, however, and it was fine.  So, there could be some kind of conflict with the theme or other gem maybe?  Or maybe there is just something stupid I'm doing.  Either way, troubleshooting may take some time with this one...
- In my test site, very large images downscaled really well, however, smaller images upscaled and the quality was not real great in my opinion.  So, there could be significant work in re-creating a good portion of images with a default large size and/or determining the standard sizes to be used.
- This would require replacing all of the current image references with a new tag that would look something like `{% responsive_image path: path/to/img.jpg %}`
- Don't believe there is any image optimization after the new images are created, so conceivably, you could end up with a smaller scaled image, but is actually more expensive to download.

## [Jekyll-picture-tag](https://github.com/robwierzbowski/jekyll-picture-tag)
### TL;DR Overview: 
Handles the solution using the `<picture>` HTML tag and it looks like that is the only option with this plugin.
### Pros:
- It does solve the requirements outlined.
- It has image optimization included which is nice
- Looks like the theme's developer has used and prefers this plugin for doing this work based on his blog.  I think this is useful information and maybe nice to know in case there are any issues getting the plugin integrated.
- Nice that it includes polyfills for browser support, though picture tag is supposed to degrade gracefully
### Cons:
- Last updated in 2015, with several listed issues/and open pull requests.  So, it doesn't appear to be all that actively maintained which to me would be a deterrent when selecting an ideal solution unless it did everything else exactly as I needed/wanted.
- Locked into a <picture> tag here it seems, so use of figures may not be possible.  If captions are required still, would need to find a solution for those.
- The use of a <picture> tag for images that don't have art direction is typically discouraged AFAIK.
- This would require replacing all of the current image references with a new tag that would look something like: `{% picture path/to/img.jpg %}`

## [Jekyll-srcset](https://github.com/netlify/jekyll-srcset)
### TL;DR Overview:
As the github repo states for this plugin it is dead simple to integrate.  However, not as configurable/flexible.  Images are resized based on pixel density as opposed to variable image sizes.  _FYI, It does appear as though there is an open bug right now and due to that, in order to get this to work, I had to use [https://github.com/niveus/jekyll-srcset](https://github.com/niveus/jekyll-srcset)._
### Pros:
- Very, very easy to setup
- has options for optimizing png images, though, not 100% sure it will work in Travis.
- Images resized based on pixel density for retina/non-retina devices.
### Cons:
- Much less flexible in terms of what image sizes you want to generate and no support for out of the box for templates.  However, we could probably create the templates on our own without an issue.
- Would require making original images optimized for 3x retina displays and then the plugin will downscale for 1x and 2x.
- Would require replacing all of the current image references with a new tag.
- Based on the fact the images are resized for pixel density as opposed to size, I'm not sure it meets the "Adapts to changes in the theme or device" requirement as it will use 1 image for non-retina devices, another for retina (@2x), and another for super retina (@3x).  However, it will use the same image for non-retina devices that have a 200px wide viewport size and non-retina devices that have a 1000px wide viewport.

## [Jekyll-cloudinary](https://github.com/nhoizey/jekyll-cloudinary)
### TL;DR Overview:
Another Jekyll plugin option that I wanted to throw out there.  I've seen a lot of folks start to use a SaaS service like this that does a lot of image optimization for sites out of the box.  I don't know if a SaaS is really something you would want to consider or even if I like the idea fully but tossing it out there as another option.
### Pros:
- Easy to implement
- solves the requirements outlined
- handles the complexity of truly responsive and optimized images for us so we don't have to worry about it
### Cons:
- Adds complexity to deployments/development, etc.
- Adds a dependency/failure point for the site.  i.e. - another third party component
- Again, would require replacing all of the current image references with a new tag.

**Comment by mtlynch** [comment]
Thanks, this is great!

I like Jekyll-responsive-image plugin best. I like that it gives us a lot of control and it seems like we can add more or less fine-grained control depending on what we want.

I've created #124 to track the migration.

Thoughts on the others:

* Jekyll-picture-tag - I'm leery of something that's unmaintained and the feature set feels fairly limiting.
* Jekyll-srcset - Doesn't seem as powerful as what I'm looking for
* Jekyll-cloudinary - This is actually pretty neat, but I run at about 90 GB bandwidth per month, so I'd have to do the $134/mo plan. Maybe their optimizations would put me under the $44/mo plan cap, but eh. I'd rather just have it all charged to Firebase.
